src/company_gui.cpp

```diff
@@ -1827,267 +1845,217 @@ struct CompanyInfrastructureWindow : Window
 		this->roadtypes &= ~_roadtypes_hidden_mask;
 	}
 
-	/** Get total infrastructure maintenance cost. */
-	Money GetTotalMaintenanceCost() const
+	/**
+	 * Build a vector of infrastructure types(roads, trains, trams, airports, stations, water vehicles) to be displayed.
+	 *   Total maintenance cost of all infrastructure types is last entry in vector.
+	 */
+	void BuildCompanyInfrastructureList() const
 	{
 		const Company *c = Company::Get((CompanyID)this->window_number);
-		Money total;
 
-		uint32_t rail_total = c->infrastructure.GetRailTotal();
-		for (RailType rt = RAILTYPE_BEGIN; rt != RAILTYPE_END; rt++) {
-			if (HasBit(this->railtypes, rt)) total += RailMaintenanceCost(rt, c->infrastructure.rail[rt], rail_total);
-		}
-		total += SignalMaintenanceCost(c->infrastructure.signal);
+		Money maintenance_cost       = 0;
+		Money total_maintenance_cost = 0;
+
+		infrastructurelist.clear();
+
+		/* railtypes + signals. */
+		infrastructurelist.emplace_back(CIW_ENTRY_HEADER, STR_COMPANY_INFRASTRUCTURE_VIEW_RAIL_SECT);
 
-		uint32_t road_total = c->infrastructure.GetRoadTotal();
-		uint32_t tram_total = c->infrastructure.GetTramTotal();
-		for (RoadType rt = ROADTYPE_BEGIN; rt != ROADTYPE_END; rt++) {
-			if (HasBit(this->roadtypes, rt)) total += RoadMaintenanceCost(rt, c->infrastructure.road[rt], RoadTypeIsRoad(rt) ? road_total : tram_total);
+		if (this->railtypes != RAILTYPES_NONE) {
+			for (const auto &rt : _sorted_railtypes) {
+				if (HasBit(this->railtypes, rt)) {
```

Comment:
Prefer early exit, `if (!HasBit(this->railtypes, rt)) continue;` Same with the others.